sdk/core/azure-core/azure/core/polling/async_base_polling.py

```diff
@@ -0,0 +1,125 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+# --------------------------------------------------------------------------
+from ..exceptions import HttpResponseError
+from .base_polling import (
+    failed,
+    BadStatus,
+    BadResponse,
+    OperationFailed,
+    LROBasePolling,
+    _raise_if_bad_http_status_and_method,
+)
+
+__all__ = ["AsyncLROBasePolling"]
+
+
+class AsyncLROBasePolling(LROBasePolling):
+    """A subclass or LROBasePolling that redefine "run" as async.
+    """
+
+    async def run(self):
+        try:
+            await self._poll()
+        except BadStatus as err:
+            self._status = "Failed"
+            raise HttpResponseError(
+                response=self._pipeline_response.http_response, error=err
+            )
+
+        except BadResponse as err:
+            self._status = "Failed"
+            raise HttpResponseError(
+                response=self._pipeline_response.http_response,
+                message=str(err),
+                error=err,
+            )
+
+        except OperationFailed as err:
+            raise HttpResponseError(
+                response=self._pipeline_response.http_response, error=err
+            )
+
+    async def _poll(self):
+        """Poll status of operation so long as operation is incomplete and
+        we have an endpoint to query.
+
+        :param callable update_cmd: The function to call to retrieve the
+         latest status of the long running operation.
+        :raises: OperationFailed if operation status 'Failed' or 'Canceled'.
+        :raises: BadStatus if response status invalid.
+        :raises: BadResponse if response invalid.
+        """
+
+        while not self.finished():
+            await self._delay()
+            await self.update_status()
+
+        if failed(self.status()):
+            raise OperationFailed("Operation failed or canceled")
+
+        final_get_url = self._operation.get_final_get_url(self._pipeline_response)
+        if final_get_url:
+            self._pipeline_response = await self.request_status(final_get_url)
+            _raise_if_bad_http_status_and_method(self._pipeline_response.http_response)
+
+    async def _sleep(self, delay):
+        await self._transport.sleep(delay)
+
+    async def _delay(self):
+        """Check for a 'retry-after' header to set timeout,
+        otherwise use configured timeout.
+        """
+        if self._pipeline_response is None:
+            return
+        response = self._pipeline_response.http_response
+        if response.headers.get("retry-after"):
+            await self._sleep(int(response.headers["retry-after"]))
```

Comment:
Retry-After may be a HTTP date. Azure services generally don't use it. But it is possible.

Reply:
ISO8601?

Reply:
I believe the format is defined [here](https://tools.ietf.org/html/rfc7231#section-7.1.1.2)

Reply:
Ok, so I will copy/paste the RFC1123 deserialzier from msrest here (don't want to depend on msrest from azure-core)

Reply:
Fixed, gosh it was not obvious, especially testing....